### PR TITLE
Disable printing by default for `DebugReducer` when running SwiftUI Previews

### PR DESCRIPTION
**Pull request by tgrapperon**
I had a case where one of my features was lagging a lot without any apparent reason in a SwiftUI Preview. The motive became immediately clear when I launched a dedicated app and saw the console filled by a very large amount of data from a `_printChanges()` that I forgot.

I've added a preliminary check to the default `_ReducerPrinter` so they don't generate any output when running SwiftUI Previews. As this only affects the default printers, this behavior can be overridden by users willing to provide their own `_ReducerPrinter` implementation (for example with variants capable of displaying this information in SwiftUI Previews like experimented in #1234).

This should make SwiftUI Previews a little snappier when a `._printChanges()` has been forgotten.

Please let me know what you think!

### CODE REVIEW

File: Sources/ComposableArchitecture/Internal/SwiftUIPreviews.swift [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import Foundation
+
+@usableFromInline
+var isRunningForPreviews: Bool {
+  #if DEBUG
+  return ProcessInfo.processInfo.environment["XCODE_RUNNING_FOR_PREVIEWS"] == "1"
+  #else
+  return false
+  #endif
+}
```
Comment (stephencelis) [inline Sources/ComposableArchitecture/Internal/SwiftUIPreviews.swift:10]: Does it work as a static check that's only performed once?

```suggestion
let isRunningForPreviews = {
  #if DEBUG
  return ProcessInfo.processInfo.environment["XCODE_RUNNING_FOR_PREVIEWS"] == "1"
  #else
  return false
  #endif
}()
```

**Comment by stephencelis** [comment]
Good to know! We typically don't use this modifier outside the app entry point, but know it's common enough for folks to do, so avoiding pitfalls like this sounds good to us!

**Review by stephencelis — APPROVED** [review]

File: Sources/ComposableArchitecture/Internal/SwiftUIPreviews.swift (hunk 1 above)
Comment (stephencelis) [reply Sources/ComposableArchitecture/Internal/SwiftUIPreviews.swift]: Actually on second thought, how about using `@Dependency(\.context)` and check for `.preview` instead? Then we could maybe even write a test for the behavior.

Comment (tgrapperon) [reply Sources/ComposableArchitecture/Internal/SwiftUIPreviews.swift]: Ok, I'll refactor

File: Tests/ComposableArchitectureTests/DebugTests.swift [hunk 2]
```diff
@@ -216,5 +216,20 @@
...
+      struct DebuggedReducer: ReducerProtocol {
+        typealias State = Int
+        typealias Action = Bool
+        func reduce(into state: inout Int, action: Bool) -> EffectTask<Bool> {
+          state += action ? 1 : -1
+          return .none
+        }
+      }
+      let store = TestStore(initialState: 0, reducer: DebuggedReducer()._printChanges())
+      store.dependencies.context = .preview
```
Comment (tgrapperon) [inline Tests/ComposableArchitectureTests/DebugTests.swift:231]: @stephencelis We can't really test the behavior directly, as there is no way to override at distance the default `TextOuputStream` from `print()` to my knowledge. Maybe there are some way to read stdout, but that seems overkill.

Switching this value to `.live` shows that something is printed, and setting it back to `.preview` shows that nothing is printed, like for the test above that runs in `.test` context.

**Comment by tgrapperon** [comment]
I refactored to use `@Dependency(\.context)` as a source of truth. This is indeed much nicer!

**Comment by iampatbrown** [comment]
Is it possible to model `_ReducerPrinter` as a dependency?

**Comment by tgrapperon** [comment]
@iampatbrown I think it was implemented as such at some point during the beta. But I'm not certain why it was rescinded. Likely because it wasn't worth the complexity at this stage.

**Comment by iampatbrown** [comment]
@tgrapperon Ah k. That makes sense :)

File: Tests/ComposableArchitectureTests/DebugTests.swift (hunk 2 above)
Comment (stephencelis) [reply Tests/ComposableArchitectureTests/DebugTests.swift:231]: Ah yeah. We should prob allow more customization in the future, but the API is underscored and barebones at the moment.

Comment (iampatbrown) [reply Tests/ComposableArchitectureTests/DebugTests.swift:231]: > there is no way to override at distance the default TextOuputStream from print()

@tgrapperon Back when you experimented with #1234 I had a go at just this :P

You can redirect stdout to `Pipe` using [dup2](https://man7.org/linux/man-pages/man2/dup.2.html)

It took a little bit of trial and error but I ended up with something that worked okay. I haven't used it though. It was just a proof of concept.

[iampatbrown/ConsoleView.swift](https://gist.github.com/iampatbrown/3034ec9756a6cdc3b9dd25d9784f3486)

Comment (tgrapperon) [reply Tests/ComposableArchitectureTests/DebugTests.swift:231]: @iampatbrown Super interesting! I'll give it a spin if I rework on #1234 idea. I'm currently using my `Logger` abstraction to duplicate prints to the UI console, but your approach could make things even simpler.

**Pull request merged**